### PR TITLE
feat(MultiEntityPicker): use EntityPresentationApi instead of humanizeEntityRef to display entity

### PR DESCRIPTION
**Pull request by Nikunj0601**
## Hey, I just made a Pull Request!

This PR is done to resolve the issue #20955 . I have changed the way to display the entity by using `entityPresentationApi` instead of `humanizeEntityRef`. I have used `entityPresentationApi` outside React jsx and `<EntityDisplayName />` inside React jsx. I have also a `promise` getter to get the promise of the entityPresenationApi for asynchronous processes of the api. Though I dont know in detail of entityPresentationApi how it is using CatalogApi, so I might be wrong. 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/scaffolder/src/components/fields/MultiEntityPicker/MultiEntityPicker.tsx [hunk 1]
```diff
@@ -64,40 +66,35 @@ export const MultiEntityPicker = (props: MultiEntityPickerProps) => {
...
 
   const catalogApi = useApi(catalogApiRef);
-
+  const entityPresentationApi = useApi(entityPresentationApiRef);
   const { value: entities, loading } = useAsync(async () => {
     const { items } = await catalogApi.getEntities(
       catalogFilter ? { filter: catalogFilter } : undefined,
     );
-    return items;
+    const primaryTitles: string[] = [];
```
Comment (vinzscam) [inline plugins/scaffolder/src/components/fields/MultiEntityPicker/MultiEntityPicker.tsx:74]: I think this change shouldn't be needed as you are now using `EntityDisplayName` for rendering the options which will invoke the entity presentation api internally 🤔

**Review by vinzscam — CHANGES_REQUESTED** [review]

File: plugins/scaffolder/src/components/fields/MultiEntityPicker/MultiEntityPicker.tsx (hunk 1 above)
Comment (Nikunj0601) [reply plugins/scaffolder/src/components/fields/MultiEntityPicker/MultiEntityPicker.tsx:74]: @vinzscam primaryTitles are added in `useAsync` to remove `getLabel` which is used in value of `Autocomplete` as it needs primaryTitle of the entities. Though if we want to completely remove `value` field of the `Autocomplete` adding primaryTitle in useAsync is no longer need. I did changes based on the requested changes in another similar PR: https://github.com/backstage/backstage/pull/24081#discussion_r1557386266

**Comment by Nikunj0601** [comment]
@freben @vinzscam for the promise getter I have added in the EntityPresentationApi, am i going in right direction or it's not what you meant?

**Comment by Nikunj0601** [comment]
@freben @vinzscam I have changed the requested changes you asked for. So I think I need to update the API report.

**Comment by Nikunj0601** [comment]
@freben @vinzscam Can you please take a look? 🙂

**Comment by Nikunj0601** [comment]
> Hi, very sorry but it seems I did some overlapping work with yours. See #24595 - I would like to keep that one if you don't mind - could you rebase on that and implement the picker in terms of that interface?

Yeah sure I will do that but can you please tell me your thoughts on changes I did regarding promise getter, like will my approach would have worked with slight modifications or was it completely off the track. Thank you your can help me understand and improve in future. 🙂

**Comment by freben** [comment]
No you were OK, no real problem, just doing things a bit differently. What happened was rather that I just had a way too big review queue and missed that this work had been done.

Main difference is that you used a getter which is good, I considered the same but left it for a future tweak if we wanted more laziness. I chose to make the promise non-optional and only to return the last result which is what usually will be what's needed. But all in all, looked okay!

**Comment by Nikunj0601** [comment]
> No you were OK, no real problem, just doing things a bit differently. What happened was rather that I just had a way too big review queue and missed that this work had been done.
> 
> Main difference is that you used a getter which is good, I considered the same but left it for a future tweak if we wanted more laziness. I chose to make the promise non-optional and only to return the last result which is what usually will be what's needed. But all in all, looked okay!

Thank you for review and feedback. I removed the changes related to API so you can review the changes done in picker.

File: plugins/scaffolder/src/components/fields/MultiEntityPicker/MultiEntityPicker.tsx [hunk 2]
```diff
@@ -64,40 +65,29 @@ export const MultiEntityPicker = (props: MultiEntityPickerProps) => {
...
       const values = refs
         .map(ref => {
           if (typeof ref !== 'string') {
             // if ref does not exist: pass 'undefined' to trigger validation for required value
-            return ref ? stringifyEntityRef(ref as Entity) : undefined;
+            return ref
+              ? entityPresentationApi.forEntity(ref, {
+                  defaultKind,
+                  defaultNamespace,
+                }).snapshot.entityRef
```
Comment (freben) [inline plugins/scaffolder/src/components/fields/MultiEntityPicker/MultiEntityPicker.tsx:89]: You can revert all of this. It just became an extremely expensive and roundabout way of getting the ref :) it'll be the exact same thing as the old code returned

File: plugins/scaffolder/src/components/fields/MultiEntityPicker/MultiEntityPicker.tsx [hunk 3]
```diff
@@ -126,7 +116,14 @@ export const MultiEntityPicker = (props: MultiEntityPickerProps) => {
...
       onChange(values);
     },
-    [onChange, formData, defaultKind, defaultNamespace, allowArbitraryValues],
+    [
+      onChange,
+      formData,
+      defaultKind,
+      defaultNamespace,
+      allowArbitraryValues,
+      entityPresentationApi,
```
Comment (freben) [inline plugins/scaffolder/src/components/fields/MultiEntityPicker/MultiEntityPicker.tsx:125]: and therefore this can be removed again too

**Review by freben — COMMENTED** [review]
Thanks, just some tweaks

**Comment by Nikunj0601** [comment]
@freben can you please look at requested changes done in the PR.

File: plugins/scaffolder/src/components/fields/MultiEntityPicker/MultiEntityPicker.tsx [hunk 4]
```diff
@@ -64,33 +65,17 @@ export const MultiEntityPicker = (props: MultiEntityPickerProps) => {
     uiSchema['ui:options']?.defaultNamespace || undefined;
 
   const catalogApi = useApi(catalogApiRef);
-
+  const entityPresentationApi = useApi(entityPresentationApiRef);
   const { value: entities, loading } = useAsync(async () => {
     const { items } = await catalogApi.getEntities(
       catalogFilter ? { filter: catalogFilter } : undefined,
     );
+
```
Comment (freben) [inline plugins/scaffolder/src/components/fields/MultiEntityPicker/MultiEntityPicker.tsx:73]: Now that https://github.com/backstage/backstage/pull/24595 is merged, could you instead do like

```ts
const entityRefToPresentation = new Map(
  await Promise.all(
    items.map(async item => {
      return [
        stringifyEntityRef(item),
        await entityPresentationApi.forEntity(item).promise
      ]
    })
  )
)

return { entities, entityRefToPresentation };
```

and return that together with the items?

Then you have a mapping from entity ref to presentation, synchronously available without having to worry about async, in all code below immediately when loading finishes

**Pull request merged**